### ISSUE TITLE
3D classification error

### ISSUE THREAD

**Issue by RubenHM80** [3dem/relion#407]
Hello,

When running 3D classification on our new workstation, the job always gets stack after right before the 1st iteration (please find attached a picture). The error is very reproducible with small or large sets of data, changing the MPI procs value ... but so far no luck. 

I'm running Relion 2.1 on a workstation using Centos 7 and 24 cores, 128 GB RAM and 4 of the new Geforce RTX 2080 GPUs as you can see in the attached image.

Please let me know If you have any suggestion and/or need more information I could provide.

Thanks,
Ruben
![3dc](https://user-images.githubusercontent.com/44416583/47397679-0a6cc700-d6f6-11e8-8ae5-1e46c1c34fd7.jpg)

**Comment by RubenHM80** [comment]
The information provided above (regarding the GPU usage) is not correct. This is how the GPUs look like when I'm running a job
![3dc2](https://user-images.githubusercontent.com/44416583/47398767-32125e00-d6fb-11e8-9128-37caf8ce1903.jpg)

**Comment by bforsbe** [comment]
Please provide 

1. the output given at the start of the run, where the GPU-mapping is given. 
2. the command used to run
3. the gpu usage as displayed by nvidia-smi, where process info is also given

May I ask that you post text, rather than pics. Thanks!

**Comment by RubenHM80** [comment]
Thanks Björn:
1) the output given at the start of the run, where the GPU-mapping is given.

=== RELION MPI setup ===
 + Number of MPI processes             = 5
 + Number of threads per MPI process  = 4
 + Total number of threads therefore  = 20
 + Master  (0) runs on host            = lxem01.sgc.loc
 + Slave     1 runs on host            = lxem01.sgc.loc
 + Slave     2 runs on host            = lxem01.sgc.loc
 + Slave     3 runs on host            = lxem01.sgc.loc
 + Slave     4 runs on host            = lxem01.sgc.loc
 =================
 uniqueHost lxem01.sgc.loc has 4 ranks.
GPU-ids not specified for this rank, threads will automatically be mapped to available devices.
 Thread 0 on slave 1 mapped to device 0
 Thread 1 on slave 1 mapped to device 0
 Thread 2 on slave 1 mapped to device 0
 Thread 3 on slave 1 mapped to device 0
GPU-ids not specified for this rank, threads will automatically be mapped to available devices.
 Thread 0 on slave 2 mapped to device 1
 Thread 1 on slave 2 mapped to device 1
 Thread 2 on slave 2 mapped to device 1
 Thread 3 on slave 2 mapped to device 1
GPU-ids not specified for this rank, threads will automatically be mapped to available devices.
 Thread 0 on slave 3 mapped to device 2
 Thread 1 on slave 3 mapped to device 2
 Thread 2 on slave 3 mapped to device 2
 Thread 3 on slave 3 mapped to device 2
GPU-ids not specified for this rank, threads will automatically be mapped to available devices.
 Thread 0 on slave 4 mapped to device 3
 Thread 1 on slave 4 mapped to device 3
 Thread 2 on slave 4 mapped to device 3
 Thread 3 on slave 4 mapped to device 3
 Running CPU instructions in double precision. 
 + On host lxem01.sgc.loc: free scratch space = 3724 Gb.
 Copying particles to scratch directory: /scratch/relion_volatile/

2) the command used to run

I run the job from the GUI. The parameters used are: 
Number of MPI procs: 5
Number of threads: 4
Submit to queue? No

I modified the # of MPI and threads without luck so far. Is that info what you need?

3) the gpu usage as displayed by nvidia-smi, where process info is also given

Wed Oct 24 09:47:36 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.57                 Driver Version: 410.57                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 2080    Off  | 00000000:17:00.0 Off |                  N/A |
|  0%   46C    P2    61W / 225W |   7222MiB /  7952MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce RTX 2080    Off  | 00000000:18:00.0 Off |                  N/A |
|  0%   49C    P2    64W / 225W |   7222MiB /  7952MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce RTX 2080    Off  | 00000000:65:00.0  On |                  N/A |
| 36%   57C    P2    64W / 225W |   7223MiB /  7951MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce RTX 2080    Off  | 00000000:B3:00.0 Off |                  N/A |
|  0%   46C    P2    63W / 225W |   7222MiB /  7952MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     46100      C   /usr/local/relion/bin/relion_refine_mpi     7211MiB |
|    1     46101      C   /usr/local/relion/bin/relion_refine_mpi     7211MiB |
|    2      1679      G   /usr/bin/X                                   276MiB |
|    2      3110      G   /usr/bin/gnome-shell                         339MiB |
|    2     46102      C   /usr/local/relion/bin/relion_refine_mpi     6603MiB |
|    3     46103      C   /usr/local/relion/bin/relion_refine_mpi     7211MiB

**Comment by RubenHM80** [comment]
2) The command line is:

*** The command is:
`which relion_refine_mpi` --continue Class3D/job013/run_it025_optimiser.star --o Class3D/job063/run_ct25 --dont_combine_weights_via_disc --no_parallel_disc_io --scratch_dir /scratch --pool 3 --iter 10 --tau2_fudge 4 --particle_diameter 350 --oversampling 1 --healpix_order 2 --offset_range 5 --offset_step 2 --j 4 --gpu ""

**Comment by bforsbe** [comment]
Looks like you are failing on a continue. Did the original run fail? What is your box-size?

**Comment by RubenHM80** [comment]
Yes, I never could run this job in this computer. 

The box size is 384 pixels. I did run the same job with the same dataset in a similar workstation (in other institute) and it worked perfectly, for that reason I can't understand what is going on.

**Comment by RubenHM80** [comment]
An additional information:, if I don't use GPU acceleration ("compute" column), the job enters into the iterations job, so I guess is something related with the GPUs

**Comment by bforsbe** [comment]
Relion considers the GPU memory in three steps (in the following order): 

1. Allocate "static" space for the reconstructions and references (fixed-size objects)
2. Reserve some "dynamic" allocation space that any process (including relion) can allocate through CUDA-allocations
3. Allocate the rest as one big "custom" allocation that is only accessible to relion - fast space that bypasses CUDA-allocations.

Relion tries to reserve 200MB in "dynamic" allocation space for each thread, that it can use for unforseeable memory need. You are using 4 threads so relion tries to keep 800MB. Thats' the 800MB referenced in your original post. 

The indication from your run is that the "static" allocation is leaving only 156MB for both dynamic and custom allocation. I'm not surprised that Relion is unable to make this work. What DOES surprise me is that there's only 156MB left. Lower resolution means low-pass (cropped) objects, requiring LESS memory. In your case relion states 60Å for a 384 box, that should NOT take 8GB of memory. 

In summary; I don't know what your issue is. Relion tends to show these warnings when you have other processes already occupying the GPU. Your original crash could be due to this, but I don't why you would see the original warning at low resolution. 
 
I recently got some 2080s to run on. It may be something funky about them, I'll let you know if that's the case.

**Comment by RubenHM80** [comment]
Thanks for the details Björn:

I tried to run the same job: this time I started with even a larger dataset (of the same particle under study) because I extracted the particles using a box-size of 280 (instead 384). Now the 3D classification is running ok.

[0] GeForce RTX 2080 | 80'C,  **66 %** |  7220 /  7952 MB | ruh(7209M)
[1] GeForce RTX 2080 | 85'C,  **71 %** |  7220 /  7952 MB | ruh(7209M)
[2] GeForce RTX 2080 | 87'C,  **74 %** |  7266 /  7951 MB | ruh(6581M) root(279M) ruh(395M) ruh(6M)
[3] GeForce RTX 2080 | 80'C,  **69 %** |  7220 /  7952 MB | ruh(7209M)


As you can see in the stats, now the cards are working at around 70% of memory, while before using a bigger box size they were at almost 100%. So i guess everything is a memory problem.

What do you think? and if you think it is a memory problem, do you think it would help to use the 11Gb version of the same cards? 

Thanks!

**Comment by RubenHM80** [comment]
Well, I thought it was going well, but every iteration is taking several hours...

**Comment by RubenHM80** [comment]
Hi Björn,

Did you find anything funky with your new 2080 cards? I'm still having the same issue running the 3D classifications, it gets stall before to start the first iteration.

Thanks

**Comment by bforsbe** [comment]
Haven't had time to check. I'll let you know when I do!

**Comment by casesongx** [comment]
thanks

**Comment by biochem-fan** [comment]
Too old issue.
